src/util/bip341.rs

```diff
@@ -0,0 +1,556 @@
+    hash_amounts: Option<sha256::Hash>,
+    /// Hash of all the prevout scriptpubkeys, computed as required
+    hash_scriptpubkeys: Option<sha256::Hash>,
+}
+
+impl<'a> SigHashCache<'a> {
+    /// Compute the sighash components from an unsigned transaction and auxiliary
+    /// in a lazy manner when required.
+    /// For the generated sighashes to be valid, no fields in the transaction may change except for
+    /// script_sig and witnesses.
+    pub fn new(tx: &'a Transaction) -> Self {
+        SigHashCache {
+            tx,
+            hash_prevouts: None,
+            hash_sequence: None,
+            hash_outputs: None,
+            hash_amounts: None,
+            hash_scriptpubkeys: None,
+        }
+    }
+
+    /// Encode the BIP341 signing data for any flag type into a given object implementing a
+    /// std::io::Write trait.
+    pub fn encode_signing_data_to<Write: io::Write>(
+        &mut self,
+        mut writer: Write,
+        input_index: usize,
+        prevouts: &Prevouts,
+        annex: Option<Vec<u8>>,
+        script_path: Option<ScriptPath>,
+        sighash_type: SigHashType,
+    ) -> Result<(), Error> {
+        if let Prevouts::All(prevouts) = prevouts {
+            if prevouts.len() != self.tx.input.len() {
+                return Err(Error::PrevoutsSize);
+            }
+        }
+
+        let (sighash, anyone_can_pay) = sighash_type.split_anyonecanpay_flag();
+
+        // epoch
+        0u8.consensus_encode(&mut writer)?;
+
+        // * Control:
+        // hash_type (1).
+        (sighash_type as u8).consensus_encode(&mut writer)?;
+
+        // * Transaction Data:
+        // nVersion (4): the nVersion of the transaction.
+        self.tx.version.consensus_encode(&mut writer)?;
+
+        // nLockTime (4): the nLockTime of the transaction.
+        self.tx.lock_time.consensus_encode(&mut writer)?;
+
+        // If the hash_type & 0x80 does not equal SIGHASH_ANYONECANPAY:
+        //     sha_prevouts (32): the SHA256 of the serialization of all input outpoints.
+        //     sha_amounts (32): the SHA256 of the serialization of all spent output amounts.
+        //     sha_scriptpubkeys (32): the SHA256 of the serialization of all spent output scriptPubKeys.
+        //     sha_sequences (32): the SHA256 of the serialization of all input nSequence.
+        if !anyone_can_pay {
+            self.hash_prevouts().consensus_encode(&mut writer)?;
+            self.hash_amounts(prevouts.get_all()?)
+                .consensus_encode(&mut writer)?;
+            self.hash_scriptpubkeys(prevouts.get_all()?)
+                .consensus_encode(&mut writer)?;
+            self.hash_sequence().consensus_encode(&mut writer)?;
+        }
+
+        // If hash_type & 3 does not equal SIGHASH_NONE or SIGHASH_SINGLE:
+        //     sha_outputs (32): the SHA256 of the serialization of all outputs in CTxOut format.
+        if let SigHashType::All | SigHashType::AllPlusAnyoneCanPay = sighash {
```

Comment:
The code as is will fail after we add `SigHashType::Default`.

Reply:
what do you suggest?

Reply:
This is just a note, that we will need to update the implementation. 

If we are going to merge this before we have `SighashType::Default`, we can follow the spec

`If hash_type & 3 does not equal SIGHASH_NONE or SIGHASH_SINGLE:`


If we are going to merge this after we have `SighashType::Default`, then we can do it as 
```
SigHashType::All | SigHashType::AllPlusAnyoneCanPay | SigHashType::Default
```

We can do the first suggestion for now.

Reply:
done in 2c4f52c